### PR TITLE
Expose prometheus agent sharding strategy

### PR DESCRIPTION
**Pull request by QuentinBisson**
Towards https://github.com/giantswarm/giantswarm/issues/30673

This PR exposes the prometheus agent sharding strategy to the helm chart so we can experiment with the scaling up and down the number of shards on some management clusters without having to release a new version to PMO.

This also needs to be added to the observability-operator

## Checklist

I have:

- [x] Described why this change is being introduced
- [x] Separated out refactoring/reformatting in a dedicated PR
- [x] Updated changelog in `CHANGELOG.md`

### CODE REVIEW

File: service/controller/resource/monitoring/remotewriteconfig/resource.go [hunk 1]
```diff
@@ -150,22 +156,22 @@ func (r *Resource) desiredConfigMap(ctx context.Context, cluster metav1.Object,
 }
 
 // We want to compute the number of shards based on the number of nodes.
-func (r *Resource) getShardsCountForCluster(ctx context.Context, cluster metav1.Object, currentShardCount int) (int, error) {
+func (r *Resource) getShardsCountForCluster(cluster metav1.Object, currentShardCount int) (int, error) {
```
Comment (QuentinBisson) [inline service/controller/resource/monitoring/remotewriteconfig/resource.go:159]: Removed useless context

File: service/controller/resource/monitoring/remotewriteconfig/shards.go [hunk 2]
```diff
@@ -1,28 +0,0 @@
...
-import "math"
-
-const (
-	shardStep                = float64(1_000_000)
-	shardScaleDownPercentage = float64(0.20)
-	shardScaleDownThreshold  = shardScaleDownPercentage * shardStep
-)
-
-// We want to start with 1 prometheus-agent for each 1M time series with a scale down 25% threshold.
-func computeShards(currentShardCount int, timeSeries float64) int {
```
Comment (QuentinBisson) [inline service/controller/resource/monitoring/remotewriteconfig/shards.go:12]: Moved that into the new type.go file

File: main.go [hunk 3]
```diff
@@ -134,6 +134,9 @@ func mainE(ctx context.Context) error {
 	daemonCommand.PersistentFlags().String(f.Service.Prometheus.ImageRepository, "giantswarm/prometheus", "Prometheus container image repository.")
 	daemonCommand.PersistentFlags().String(f.Service.Prometheus.Version, "v2.47.1", "Prometheus container image version.")
 
+	daemonCommand.PersistentFlags().Float64(f.Service.PrometheusAgent.ShardScaleUpSeriesCount, 1_000_000, "Prometheus agent shard scale up series count to know when to add a shard.")
+	daemonCommand.PersistentFlags().Float64(f.Service.PrometheusAgent.ShardScaleDownPercentage, 0.20, "Prometheus agent shard scale down percentage to know when the to scale down the number of shards.")
```
Comment (marieroque) [inline main.go:138]:
```suggestion
	daemonCommand.PersistentFlags().Float64(f.Service.PrometheusAgent.ShardScaleDownPercentage, 0.20, "Prometheus agent shard scale down percentage to know when to scale down the number of shards.")
```

File: service/controller/resource/monitoring/remotewriteconfig/types.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import "math"
+
+type PrometheusAgentShardingStrategy struct {
+	// Configures the number of series needed to add a new shard. Computation is number of series / ShardScaleUpSeriesCount
+	ShardScaleUpSeriesCount float64
+	// Percentage of needed series based on ShardScaleUpSeriesCount to scale down agents
+	ShardScaleDownPercentage float64
+}
+
+// We want to start with 1 prometheus-agent for each 1M time series with a scale down 25% threshold.
```
Comment (marieroque) [inline service/controller/resource/monitoring/remotewriteconfig/types.go:12]:
```suggestion
// We want to start with 1 prometheus-agent for each 1M time series with a scale down 20% threshold.
```

**Review by marieroque — APPROVED** [review]
LGTM

File: helm/prometheus-meta-operator/templates/configmap.yaml [hunk 5]
```diff
@@ -41,6 +41,15 @@ data:
         {{- if .Values.prometheus.version }}
         version: {{ .Values.prometheus.version }}
         {{- end }}
+      {{- if (.Values.prometheusAgent).shardScaleUpSeriesCount (.Values.prometheusAgent).shardScaleDownPercentage }}
```
Comment (hervenicol) [inline helm/prometheus-meta-operator/templates/configmap.yaml:44]: What is this syntax?
I see parenthesis and multiple conditions but no operator between the different conditions, and I can imagine what it does but I can't be sure.

Comment (QuentinBisson) [reply helm/prometheus-meta-operator/templates/configmap.yaml:44]: This is the helm syntax to say that .Values.prometheusAgent can be nil :)

Comment (QuentinBisson) [reply helm/prometheus-meta-operator/templates/configmap.yaml:44]: This actually relies on the order of operations so it executes the inner loop before it executes the other one

Comment (hervenicol) [reply helm/prometheus-meta-operator/templates/configmap.yaml]: So it's like `{{- if .Values.prometheusAgent AND .Values.prometheusAgent.shardScaleUpSeriesCount AND .Values.prometheusAgent AND .Values.prometheusAgent.shardScaleDownPercentage }}` ?
I didn't know you could put several conditions in one `if` with no operator between.

Comment (hervenicol) [reply helm/prometheus-meta-operator/templates/configmap.yaml]: Oh, I guess that answers my question :rofl: 
image

**Review by hervenicol — APPROVED** [review]
LGTM

**Comment by QuentinBisson** [comment]
Tested on gaia

**Pull request merged**